internal/transport/http2_server.go

```diff
@@ -933,9 +932,14 @@ func (t *http2Server) checkForHeaderListSize(it interface{}) bool {
 
 // WriteHeader sends the header metadata md back to the client.
 func (t *http2Server) WriteHeader(s *Stream, md metadata.MD) error {
-	if s.updateHeaderSent() || s.getState() == streamDone {
+	if s.getState() == streamDone {
+		return ContextErr(s.ctx.Err())
```

Comment:
There's a small window where the state can be `streamDone` but the context hasn't been canceled yet.  In these cases, we would be returning an INTERNAL error.

We could paper over this race by calling `s.cancel()` here, as we seem to do in `Write`.  But honestly I think the right fix here is to move `s.cancel()` in `deleteStream` to the first line of both `finishStream` and `cancelStream`.  And then we can remove the `s.cancel()` in `Write`.  Would you mind doing this and hopefully (:crossed_fingers:) all the tests still pass?

Looking closer at `Write`, it looks like it has special logic to prefer returning an error that the whole connection was closed over the stream being closed.  We probably want to do the same thing here.  If you want to copy/paste the select, that's fine.  Or if you don't mind refactoring, it looks like we need:

```go
func (t *http2Server) streamContextErr(s *Stream) error {
	select {
	case <-t.done:
		return ErrConnClosing
	default:
	}
	return ContextErr(s.ctx.Err())
}
```

But I can do that as a follow-on later otherwise.

Reply:
Yup, tests seem to all pass with this!

Reply:
It _is_ done by `deleteStream`, but there is a small race, still, where the context is not canceled while the state is `streamDone`.  So I was suggesting moving that cancelation to happen _before_ setting the stream's state to `streamDone`, and also removing it from `deleteStream`.